File: src/main/scala/clase1/PatternMatching_Length.scala
package clase1

object PatternMatching_Length extends App {
  val length = (list: List[Int]) => {

    def doLength(list: List[Int], acc: Int): Int = list match {
      case head :: tail => doLength(tail, acc + 1)
      case Nil => acc
    }

    doLength(list, 0)

  }

  println(length(List(1, 2, 3)))
  println(length(List(1)))
  println(length(List()))

}

File: haskell/src/clase1/PatternMatchingLength.hs
module PatternMatchingLength where

doLength (head:tail) acc = doLength tail (acc + 1)
doLength [] acc = acc

listLength list = doLength list 0


main = do
    print (listLength [1, 2, 3])
    print (listLength [1])
    print (listLength [])
